ckan_datapackage_tools/converter.py

```diff
@@ -177,82 +174,169 @@ def _parse_notes(dataset_dict):
     return result
 
 
-def _parse_license(dataset_dict):
+def _parse_tags(dataset_dict):
     result = {}
-    license = {}
 
-    if dataset_dict.get('license_id'):
-        license['type'] = dataset_dict['license_id']
-    if dataset_dict.get('license_title'):
-        license['title'] = dataset_dict['license_title']
-    if dataset_dict.get('license_url'):
-        license['url'] = dataset_dict['license_url']
+    keywords = [tag['name'] for tag in dataset_dict.get('tags', [])]
 
-    if license:
-        result['license'] = license
+    if keywords:
+        result['keywords'] = keywords
 
     return result
 
 
-def _parse_author_and_source(dataset_dict):
+def _parse_extras(dataset_dict):
     result = {}
-    source = {}
-
-    if dataset_dict.get('author'):
-        source['name'] = dataset_dict['author']
-    if dataset_dict.get('author_email'):
-        source['email'] = dataset_dict['author_email']
-    if dataset_dict.get('url'):
-        source['web'] = dataset_dict['url']
+    extras = [[extra['key'], extra['value']] for extra
+              in dataset_dict.get('extras', [])]
+    for extra in extras:
+        try:
+            extra[1] = json.loads(extra[1])
+        except (ValueError, TypeError):
+            pass
+    sources = _parse_organization_as_primary_source(dataset_dict)
+    licenses = _parse_primary_license(dataset_dict)
+    contributors = _parse_primary_contributors(dataset_dict)
+    if extras:
+        extras_dict = dict(extras)
+        result.update(_parse_profile(extras_dict))
+        _parse_extra_sources(extras_dict, sources)
+        _parse_extra_licenses(extras_dict, licenses)
+        _parse_extra_contributors(extras_dict, contributors)
+        if extras_dict:
+            result['extras'] = extras_dict
+    if sources:
+        result['sources'] = sources
+    if licenses:
+        result['licenses'] = licenses
+    if contributors:
+        result['contributors'] = contributors
+    return result
 
-    if source:
-        result['sources'] = [source]
 
+def _parse_profile(dataset_dict):
+    result = {}
+    profile = dataset_dict.pop('profile', '')
+    if profile:
+        result['profile'] = profile
     return result
 
 
-def _parse_maintainer(dataset_dict):
-    result = {}
-    author = {}
+def _parse_organization_as_primary_source(dataset_dict):
+    all_sources = []
+    source = {}
+    organization = dataset_dict.get('organization')
+    if organization.get('is_organization', False):
```

Comment:
I'm reluctant to use info from organizations on the DP output unless it can also be set on the way, when ingesting a DP. It's going to be difficult to ensure that imported DP are assigned to CKAN organizations reliably. The only thing we have to match both are the `sources` `title` property. But maybe it's worth having sources always populated even if not supporting it on the way in.

Reply:
yep same as previous comment.
The only thing I'm not sure of, though, is what to do about dataset's source url.
Because the latest frictionless states that 'title' is mandatory, but if there is an existing dataset source url, surely downloaded datapackage should be capturing this somehow (as it also has 'sources'). Or is this a bigger question for discussion with @pwalsh or @rufuspollock ?
In another, related project, [Data Curator](https://github.com/ODIQueensland/data-curator), in which we capture datapackage properties, we are looking to store properties 'silently' from incoming datasets to ensure that the properties still exist on upload again to ckan, but because 'source'(s) exists in both dataset and datapackage, I thought perhaps this might need to be handled some other way?
For that matter, as it seems important for a dataset to have an organisation, I'm not sure there's an easy way to capture that in a datapackage or whether datapackage should capture this at all (as it has major impact on how datasets are loaded into ckan)
Apologise if these are questions that @Stephen-Gates may have already squared away with you, but due to many conflicting commitments, unfortunately for us, Stephen has had to step back from his role with ODIQueensland, so doing what I can to fill in gaps here :P

Reply:
These questions were not answered because they are tricky to answer :)

I think my preference would be to only add organizations to sources in the way out for now as this is the most robust approach. Trying to match sources to organizations on the way in just using the title seems too brittle.

Reply:
Hi @amercader 
Yes I agree - perhaps we could add organisation (just the title) on the way out - but I might just leave this out for now as then there's a problem with not repeating it in 'extras' on the way in with repeated imports.
I've realised since that probably the mistake I made was thinking a CKAN dataset source is/should be the same as a datapackage source. I realise now that with the new contributors, and that author_email, author_name and source_url were all being collected together, then source_url from ckan should probably also go the contributor, with role 'author' as contributor['path'] in the datapackage.